ouroboros-network-api/src/Ouroboros/Network/AnchoredSeq.hs

```diff
@@ -807,5 +807,5 @@ filterWithStopSpec p stop = goNext []
           -> AnchoredSeq v a b
           -> AnchoredSeq v a b
     join' a b =
-        fromMaybe (error "could not join sequences") $
+        fromMaybe (error "could not join sequences") $ -- TODO
         join (\_ _ -> True) a b
```

Comment:
If desired, this error could be eliminated by creating a total/irrefutable version of `join` that glues its provided sequences together without checking their compatibility.

Reply:
@nfrisby do you think it's worth it?

Reply:
> creating a total/irrefutable version of join that glues its provided sequences together without checking their compatibility

Just for context: I would not call such a function "total", since the resulting anchored seq is potentially invalid. Along those lines, if such a function were to exist, I think its name should start with the prefix `unsafe*`.

If the new function has the `unsafe*` prefix, then I would only have a very slight preference for leaving this call site as it is compared to defining and using the new function.

---

I don't know the context within this function, but I recall an idea elsewhere to avoid an impossible case by enriching the interface of a preceding call (something like replacing `... checkIntersect ... join ...` with `... checkIntersectAndIfSoJoin ...`).

It looks like the context here is some low-level pattern matching, so maybe that's not immediately obvious. Though perhaps the joining could be happening simultaneously in a new accumulator argument, or something like that?

Reply:
> I would only have a very slight preference for leaving this call site as it is compared to defining and using the new function.

That's reasonable.

> something like replacing `... checkIntersect ... join ...` with `... checkIntersectAndIfSoJoin ...`

A function like that seems like it'd be partial as well (the implied `checkIntersectAndIfNot` case), so we'd have to figure out how to propagate that partiality through or out of this function.

> the joining could be happening simultaneously in a new accumulator argument

I guess an approach like that would eliminate the error, but it would also be indirectly duplicative of `join`, and would introduce the potential to diverge from `join`'s semantics, if they were to change.

Given all this, it seems to me like the existing `error`-based approach might be a local minimum/maximum of complexity/safety, so I'm minded to recommend that we just leave it as-is. Thoughts?

Reply:
Let's leave it as is then.